### PR TITLE
LTP: Write results to JSON file

### PR DESCRIPTION
**Pull request by richiejp**
We already do some parsing of the LTP log in OpenQA, so we can use the result
of that parsing to also write out a structured JSON file with test results in it.

Output may look something like
```
{
   "format" : "result_array:v1",
   "results" : [
      {
         "environment" : {
            "harness" : "SUSE OpenQA",
            "libc" : "",
            "kernel" : "",
            "product" : "sle:12-SP3",
            "revision" : "0344",
            "arch" : "qemu_x86_64"
         },
         "test" : {
            "duration" : 0.00689668099221308,
            "log" : "\r\nabs01       1  TPASS  :  Test passed\r\nabs01       2  TPASS  :  Test passed\r\nabs01       3  TPASS  :  Test passed\r\n### TEST abs01 COMPLETE >>> 0"
         },
         "pass" : "true",
         "test_fqn" : "LTP:math:abs01"
      },
      {
         "test" : {
            "duration" : 0.00675127700378653,
            "log" : "\r\natof01      1  TPASS  :  Test passed\r\natof01      2  TPASS  :  Test passed\r\natof01      3  TPASS  :  Test passed\r\natof01      4  TPASS  :  Test passed\r\n### TEST atof01 COMPLETE >>> 0"
         },
         "environment" : {
            "kernel" : "",
            "libc" : "",
            "harness" : "SUSE OpenQA",
            "product" : "sle:12-SP3",
            "revision" : "0344",
            "arch" : "qemu_x86_64"
         },
         "pass" : "true",
         "test_fqn" : "LTP:math:atof01"
      },
  ]
}
```

This is suitable for being dumped straight into a key-value store like Redis (which has JSON values) or a document store like CouchDB. Alternatively it could be parsed and put into an SQL database. The structure is explained in the code, line 210.

https://progress.opensuse.org/issues/19104

@mimi1vx

### CODE REVIEW

**Comment by mimi1vx** [comment]
@richiejp `"pass" : "true",` , i think this is too simple indicator ... because Test results can have more states than PASSED or FAILED ( SKIPPED/IGNORED, config error etc )

**Comment by richiejp** [comment]
@mimi1vx Thanks for feedback, I will add a more complex result in the test section, like this

```
{
         "environment" : {
            "harness" : "SUSE OpenQA",
            "libc" : "",
            "kernel" : "",
            "product" : "sle:12-SP3",
            "revision" : "0344",
            "arch" : "qemu_x86_64"
         },
         "test" : {
            "result": "TPASS"
            "duration" : 0.00689668099221308,
            "log" : "\r\nabs01       1  TPASS  :  Test passed\r\nabs01       2  TPASS  :  Test passed\r\nabs01       3  TPASS  :  Test passed\r\n### TEST abs01 COMPLETE >>> 0"
         },
         "pass" : "true",
         "test_fqn" : "LTP:math:abs01"
      },
```

I will have to rewrite quite a bit of run_ltp.pm to do this, but it seems necessary.
Also I could make it an array of objects, because LTP tests can return multiple results, so

```
"results": [{ "key": "TPASS", "value":"Test passed" }, { "key": "TWARN", "value", "Some warning!"}]
```

This could be added later as a separate field. It might be overkill, when the log is already included.

The pass field needs to be simple so that it can mean almost the same thing for all test suites. I could maybe make it ternary, that is have 'pass', 'fail' and 'null'. Where 'null' means the test neither passed nor failed, which covers TCONF or being skipped altogether. Only problem is people might interpret a test returning no result as null, but if the test is broken then that might not be helpful, as null results might just be ignored. Anyway, I can rename the pass field to 'status' and decide later.

**Comment by richiejp** [comment]
Example output: http://paste.opensuse.org/4941367

**Comment by richiejp** [comment]
@pevik fyi

File: tests/kernel/run_ltp.pm [hunk 1]
```diff
@@ -117,132 +118,203 @@ sub parse_openposix_runfile {
...
+            elsif ($1 == 32 && $results->{tfail} + $results->{tbrok}) {
+                say $fh 'TEST EXIT CODE IS 32 (TCONF), YET TFAIL OR TBROK WAS SEEN!';
             }
             elsif ($1 == 32) {
                 say $fh 'Test process returned TCONF (32).';
-                $tconf = 1;
+                $results->{tconf}++;
+            }
+            elsif ($1 == 4 && $results->{tfail} + $results->{tbrok}) {
+                say $fh 'TEST EXIT CODE IS 4 (TWARN), YET TFAIL OR TBROK WAS SEEN!';
```
Comment (pevik) [inline tests/kernel/run_ltp.pm]: I wonder, if these elsif should be checking `$results->{tpass}` as well.

**Pull request merged**